feat(oauth2): add Spring Security OAuth2 documentation (#1677)
camunda/camunda-docs-manual: content/user-guide/spring-boot-integration/spring-security.md

```xml
<dependency>
  <groupId>org.camunda.bpm.springboot</groupId>
  <artifactId>camunda-bpm-spring-boot-starter-security</artifactId>
</dependency>
```

# Camunda Run

In order to enable the Spring Security OAuth2 integration in Camunda Run, start Run with an extra `--oauth2` argument:

```shell
./start.sh --webapps --rest --example --oauth2
```

# Auto Configuration

The Camunda integration has two default auto configurations. Depending on the OAuth2 client
registration in the application properties (`spring.security.oauth2.client.registration`) either the
`CamundaSpringSecurityOAuth2AutoConfiguration` or the `CamundaBpmSpringSecurityDisableAutoConfiguration` will be used.

## OAuth2 Auto Configuration

Starts if there is OAuth2 client registration configured. This class configures the Spring
Security filter chain to a permit all mode.

Spring auto configuration class: {{< javadocref page="org/camunda/bpm/spring/boot/starter/security/oauth2/impl/CamundaSpringSecurityOAuth2AutoConfiguration.html" text="CamundaSpringSecurityOAuth2AutoConfiguration" >}}

## Disable Auto Configuration

> Starts if there is **no** OAuth2 client registration configured. This class configures the Spring

Author: Changed to _Configuration activates_

For the client registration, please refer to the official Spring Security's [OAuth2 Core Configuration][OAuth2Config] documentation to configure your choice of identity provider.

Camunda's integration uses the **name** field from Spring Security's principal object for the user
authentication. This needs to match the User ID in Camunda.

{{< note title="Heads Up!" class="info" >}}
At this stage, OAuth2 is only used for authentication. Meaning, that the user needs to be
also configured with the matching User ID and proper authorizations in Camunda.

If the user is not available or doesn't have sufficient authorizations, they won't be able to access the Webapps.
{{< /note >}}

### User Name Mapping

Spring Security provides an attribute to override the default scope (`sub`) used for the username.
This is the `spring.security.oauth2.client.provider.[providerId].user-name-attribute` from the
above-mentioned [OAuth2 Core Configuration][OAuth2Config].

If you wish to use a default attribute (claim) as the username, then make sure to override the `user-name-attribute` in your application properties.

### Initial User

> In order to create an initial user in your application, you can either use the `camunda.bpm.admin-user` property:

Author: I added a bit more info, let me know what you think.

```yaml
camunda.bpm:
  admin-user:
    id: demo
    password: demo
    firstName: Demo
    lastName: Demo
```

Or the `AdministratorAuthorizationPlugin`, see [The Administrator Authorization Plugin]({{< ref "/user-guide/process-engine/authorization-service.md#the-administrator-authorization-plugin" >}}) documentation for more details. 

## OAuth2 Identity Provider

Additionally to the OAuth2 login, Camunda also provides support to use groups from OAuth2.
This is achieved with a custom [identity service]({{< ref "/user-guide/process-engine/identity-service.md" >}}), called {{< javadocref page="org/camunda/bpm/spring/boot/starter/security/oauth2/impl/OAuth2IdentityProvider.html" text="OAuth2IdentityProvider" >}}

This is a read-only identity provider that configures user's groups from the [Spring Security's granted authorities][Authorities].
Additionally, the identity provider also contains the default Camunda Database Identity Service as a fallback.

TODO add more details

In order to activate this, add the following properties:
```yaml

> camunda.bpm.oauth2:

Reviewer: ❓ We need to add the properties to the [configurations table](https://github.com/camunda/camunda-docs-manual/blob/4454-oauth2-docs/content/user-guide/spring-boot-integration/configuration.md#camunda-engine-properties) or we don't want to advertise the feature there?

Author: I added configuration sections where needed (identity provide & logout)

Reviewer: ❓ Does this mean that they won't be mentioned in the default configuration page of Spring boot? I don't see changes there.

Author: Ah good point, I overlooked it, I will add them there too.

  identity-provider:
    enabled: true
```

### Authorities mapping

We also provide a default granted authorities mapper, that can override the Spring Security
authorities, that are populated by default with the scope claim.

This mapper can be enabled with the `camunda.bpm.oauth2.identity-provider.group-name-attribute` property:
```yaml
camunda.bpm.oauth2:
  identity-provider:
    enabled: true
    group-name-attribute: cognito:groups
```

Alternatively, you can also define your own [GrantedAuthoritiesMapper][GrantedAuthoritiesMapper], if you need more customization.

## Security Recommendations

If you decide to use OAuth2 for login in Camunda, we highly recommend to consult and implement the current industry recommended security standards.

> Additionally, also follow the security recommendations specified by your identity provider.

Reviewer: ✍️ I think similar thing (with link to this page?) must be mentioned here as well: https://docs.camunda.org/manual/latest/user-guide/security/#security-configuration-in-the-external-environment
(Or somewhere else on that page)

Reviewer: We can iterate on the docs after we receive user's feedback. As long as all configuration must be done on identity provider's side, we don't need to list explicitly anything.. You already added the token lifetime for which we can anticipate to be asked for. So we are good.

### Token Lifetime

As OAuth2 works with the exchange of tokens and tokens are valid until the specified expiration (
`exp` claim), it is inevitable that in a few cases tokens might outlive the main SSO session.
Meaning, the user might be already logged out but still have valid tokens on other pages.

In order to minimize the occurrence if this, we recommend the use of short-lived access tokens along with refresh tokens.

TODO add some links
? https://datatracker.ietf.org/doc/html/draft-ietf-oauth-security-topics

> ## Disabling Auto Configuration

Reviewer: 📓 This needs to be manually tested.

Author: Test both in Spring Boot, they work.

Reviewer: I forgot that I wrote that and I was about to ask again. 😄

weight: 80

menu:
  main:
    identifier: "user-guide-spring-security-integration"
    parent: "user-guide-spring-boot-integration"

---

Camunda provides Spring Security OAuth2 integration with the `camunda-bpm-spring-boot-starter-security` library.
This library contains the Spring Security and Spring Security OAuth2 dependencies along with
configuration classes that integrate Spring Security with Camunda Webapp's authentication.

This is available both for Spring Boot and Camunda Run.

Camunda's integration comes with multiple components and configurations. In the next sections you can find more details to each of them.

# Activate OAuth2

## Spring Boot

In order to enable the Spring Security OAuth2 integration in Spring Boot, add the following dependency to your project:

```xml
<dependency>
  <groupId>org.camunda.bpm.springboot</groupId>
  <artifactId>camunda-bpm-spring-boot-starter-security</artifactId>
</dependency>
```

## Camunda Run

Camunda Run already contains the required libraries, all you need to do is to activate them.
In order to enable the Spring Security OAuth2 integration in Camunda Run, start Run with an extra `--oauth2` argument:

```shell

> ./start.sh --webapps --rest --example --oauth2

Reviewer: 🔧 The example is not recommended for production so I suggest to remove it:
```suggestion
./start.sh --webapps --rest --oauth2
```

Security's [OAuth2 Core Configuration][OAuth2Config] documentation to configure your choice of
identity provider.

Once there is an OAuth2 client registration configured and the Camunda Spring Security OAuth2
integration is enabled, the Webapps will use the configured OAuth2 provider for
authentication.

# User Name Mapping

Camunda's integration uses the **name** field from Spring Security's principal object as the User ID
in the Webapps.

Spring Security by default uses the subject (`sub`) claim as the principal name. As the User ID in Camunda is
an important part for authorizations, it's important that the right claim is used. 

Spring Security provides a way to change the default attribute used for the username.
This is the `spring.security.oauth2.client.provider.[providerId].user-name-attribute` from the
above-mentioned [OAuth2 Core Configuration][OAuth2Config].

{{< note title="Heads Up!" class="info" >}}
Make sure to correctly configure which token attribute should be used as the User ID.
{{< /note >}}

## Configuring Initial Authorizations

For creating initial authorizations in your application, you have the following options available:

> 1. The `camunda.bpm.admin-user` property to create an admin user with authorizations:

Reviewer: 🙃 
```suggestion
1. The `camunda.bpm.admin-user` property to create an administrator user with authorizations:
```

   ```yaml
   camunda.bpm:
     admin-user:
       id: demo
       password: demo
       firstName: Demo
       lastName: Demo
   ```

  - See [Camunda Engine Properties]({{< ref "/user-guide/spring-boot-integration/configuration#camunda-engine-properties" >}}) documentation for more details.
2. The [Administrator Authorization Plugin]({{< ref "/user-guide/process-engine/authorization-service.md#the-administrator-authorization-plugin" >}})
to grant administrator authorizations for a particular OAuth2 user or group.

# OAuth2 Identity Provider

Additionally to the OAuth2 login, Camunda also provides support to use groups from OAuth2.

> This is achieved with a custom [identity service]({{< ref "/user-guide/process-engine/identity-service.md" >}}), called {{< javadocref page="org/camunda/bpm/spring/boot/starter/security/oauth2/impl/OAuth2IdentityProvider.html" text="OAuth2IdentityProvider" >}}

Reviewer: 🔧 Missing fullstop:
```suggestion
This is achieved with a custom [identity service]({{< ref "/user-guide/process-engine/identity-service.md" >}}), called {{< javadocref page="org/camunda/bpm/spring/boot/starter/security/oauth2/impl/OAuth2IdentityProvider.html" text="OAuth2IdentityProvider" >}}.
```

This is a read-only identity provider that configures user's groups from the [Spring Security's granted authorities][Authorities].
This identity provider also supports the default Camunda Database Identity Service as a fallback for authentications for the REST API.

The identity provider is activated by default. You can override this with the following properties:
```yaml
camunda.bpm.oauth2:
  identity-provider:
    enabled: false
```

> See [Configuration](#configuration) section for more infos.

Reviewer: 🙃 
```suggestion
See [Configuration](#configuration) section for more information.
```

## Granted Authorities Mapper

We also provide a default granted authorities mapper, that can override the Spring Security
authorities, that are by default populated with the scope (`scp`) claim.

This mapper can be enabled with the `group-name-attribute` property:
```yaml
camunda.bpm.oauth2:
  identity-provider:
    enabled: true
    group-name-attribute: cognito:groups
```

The mapper is only activated if the property is configured.
It supports claims with types of collection of `String`s and `String`.
If the claim is a `String`, it will try to split it with a delimiter which is comma by default.
You can override the default delimiter with the `group-name-delimiter` property.

> See [Configuration](#configuration) section for more infos.

Reviewer: 🙃 
```suggestion
See [Configuration](#configuration) section for more details.
```

Author: Changed to _information_.

## Disabling Identity Provider

With the [above-mentioned property](#configuration), the identity provider can be deactivated.
Without identity provider OAuth2 is only used for authentication. This means, that the user needs to
be also configured with the matching User ID in Camunda database.

If the user is not available or doesn't have sufficient authorizations, they won't be able to access
the Webapps.

# Logout

We provide support for local and client initiated SSO logout as well.
In order to support both logouts, the Camunda integration also contains a Frontend Plugin that overrides the Webapps default logout behaviour.
As a consequence, when the Webapp user clicks on the logout, it invokes Spring's logout endpoint (`/logout`) instead of Camunda's.

## Client Initiated SSO Logout

We support client initiated OIDC SSO logout.
Please refer Spring's [OpenID Connect 1.0 Client-Initiated Logout][SSOLogout] section for more information.

In order to configure this feature, use the following properties:

```yaml
camunda.bpm.oauth2:
  sso-logout:
    enabled: true
    postLogoutRedirectUri: https://camunda.com/
```

> See [Configuration](#configuration-1) section for more infos.

Reviewer: ❌ That linking is not working locally for me. I suggest to remove the reference to avoid confusion.
```suggestion
Check the next section for more information.
```

Author: Weird, it's working for me.

    <td>Configures the URI the user is redirected after SSO logout from the provider.</td>
    <td><code>{baseUrl}</code></td>
  </tr>
</table>

## Limitations

Currently, it's not possible to change the default Spring logout endpoint, which is `/logout`.

# Security Recommendations

Camunda's integration heavily relies on Spring Security's OAuth2 support.

If you decide to use OAuth2 for login in Camunda, we highly recommend to consult and implement the current industry recommended security standards.
Additionally, also follow the security recommendations specified by your identity provider.

## Token Lifetime

As OAuth2 works with the exchange of tokens and tokens are valid until the specified expiration (`exp`),
it is inevitable that in a few cases tokens might outlive the main SSO session.
Meaning, the user might be already logged out but still have valid tokens on other pages.

In order to minimize the risk of this, we recommend the use of short-lived access tokens along with
refresh tokens.
Refresh tokens can be revoked, and issuing new access tokens require interaction with the provider,
which means the user session can be revalidated more frequently.

# Logging

> We use debug logging for several things such as bean registrations, user authentication or logout and token authorization.

Reviewer: 🔧 
```suggestion
You can switch the level of the following logger to track: bean registrations, user authentication or logout, and token authorization.
```

Author: Applied suggestion.

an important part for authorizations, it's important that the right claim is used. 

Spring Security provides a way to change the default attribute used for the username.
This is the `spring.security.oauth2.client.provider.[providerId].user-name-attribute` from the
above-mentioned [OAuth2 Core Configuration][OAuth2Config].

{{< note title="Heads Up!" class="info" >}}
Make sure to correctly configure which token attribute should be used as the User ID.
{{< /note >}}

## Configuring Initial Authorizations

In order to create initial authorizations in your application, you can either use the `camunda.bpm.admin-user` property to create an admin user with authorizations:

```yaml
camunda.bpm:
  admin-user:
    id: demo
    password: demo
    firstName: Demo
    lastName: Demo
```

See [Camunda Engine Properties]({{< ref "/user-guide/spring-boot-integration/configuration#camunda-engine-properties" >}}) documentation for more details.

Or the [Administrator Authorization Plugin]({{< ref "/user-guide/process-engine/authorization-service.md#the-administrator-authorization-plugin" >}})

> to grant administrator authorizations for a particular OAuth2 user or group.

Reviewer: 🔧 The two options are too far away for me. Could you consider bullets or rewording?

Author: Added bullet points.

Additionally, also follow the security recommendations specified by your identity provider.

## Token Lifetime

As OAuth2 works with the exchange of tokens and tokens are valid until the specified expiration (`exp`),
it is inevitable that in a few cases tokens might outlive the main SSO session.
Meaning, the user might be already logged out but still have valid tokens on other pages.

In order to minimize the risk of this, we recommend the use of short-lived access tokens along with
refresh tokens.
Refresh tokens can be revoked, and issuing new access tokens require interaction with the provider,
which means the user session can be revalidated more frequently.

# Logging

We use debug logging for several things such as bean registrations, user authentication or logout and token authorization.
Logging can be enabled for the package via the following property:

```yaml
logging:
  level:
    org.camunda.bpm.spring.boot.starter.security.oauth2: DEBUG
```

# Example

In this section we provide an example configuration with OKTA as OIDC provider.

> Additionally, we also mark and explain a few lines:

Author: 💬 I also added a full example config with some hints. Let me know if you find it unnecessary.
